### PR TITLE
[WIP] Allow foo.'bar-bat' and foo.1 to access object members

### LINKED ISSUE #3520: Cleaner syntax for accessing "illegal" object members: object.'some-value'
This comes up a lot when working with compiled templates in SPAs and I imagine in other places as well. We end up with template objects that have a lot of members who's names are not legal JavaScript identifiers because they are based on the original file names of the templates.  Needing to wrap brackets and quotes around the member names seems like more syntax that is needed.

Would it be possible to make:

```
templates.'my-template-name'
#and
templates."another-template-name"
```

valid syntax equivalent to:

```
templates.['my-template-name']
#and
templates.["another-template-name"]
```

### PR DESCRIPTION
**Pull request by pscollins**
This change addresses  #3520. I wasn't sure if there is a consensus on that issue, but I think that it's a simple change that could clean up code in some places.

### CODE REVIEW

**Comment by epidemian** [comment]
:+1: 

This makes for a nice symmetry with literal object notation.

```coffeescript
obj = 
  # Can use barewords for keys that look like identifiers.
  foo: 42
  # But need enclosing commas for non-identifier-like keys.
  'wordy key': 43

# Can access identifier-like keys by a bareword after the dot.
obj.foo # => 42
# But need enclosing commas for accessing non-identifier-like keys.
obj.'wordy key' # => 43
```

**Comment by jashkenas** [comment]
I'm not sure really what to think about this. There is a certain symmetry, but there's also a certain awkwardness, especially when it comes to more complicated accesses:

    configuration.'ajax-get-request'(url, options)

... that's just odd. Any staunch defenders of this idea?

**Comment by vendethiel** [comment]
It works in coco/LS, but I admit I don't use it that much. It's only interesting if you have to deal with "strange" json schemas...

**Comment by lydell** [comment]
_If_ this is going to be supported, numbers should be as well. `a.1`

**Comment by michaelficarra** [comment]
@lydell: Agreed. I'm :+1: for both of them.

**Comment by pscollins** [comment]
@lydell maybe it would be easiest to do something like

    PropertyIdentifier ::= Identifier | AlphaNumeric 

and then replacing the existing uses of `Identifier` with `PropertyIdentifier` in `Accessor`?

**Comment by pscollins** [comment]
@lydell maybe it would be easiest to do something like

    PropertyIdentifier ::= Identifier | AlphaNumeric 

and then replacing the existing uses of `Identifier` with `PropertyIdentifier` in `Accessor`?

**Comment by lydell** [comment]
> especially when it comes to more complicated accesses:
> 
>     configuration.'ajax-get-request'(url, options)
> 
> ... that's just odd.

I always try to avoid the `a[b](c)`, because IMO it is not very readable. `configuration.'ajax-get-request'(url, options)` isn’t more odd than `configuration['ajax-get-request'](url, options)` or the syntactically valid JS `"string"()`.

**Comment by jashkenas** [comment]
Alright then. I'm persuaded. Let's add numbers to this patch, and some more robust tests, and merge.

**Comment by michaelficarra** [comment]
:grinning:

**Comment by lydell** [comment]
@pscollins Please do what you suggested and add tests for all the access variants (not just `.`) and tests that you can call such accesses as well as access deeper properties.

**Comment by pscollins** [comment]
@lydell `::` and `?::` as well, or just `.` and `?.`? My guess is no, since things accessible through `::` won't have been written as object literals, and so shouldn't have crazy names.

**Comment by michaelficarra** [comment]
@pscollins: Of course those should be allowed.

```coffee
a::' ' = ->
a.prototype = ' ': ->
```

**Comment by jashkenas** [comment]
Yes. If it's for any accessor, it should be for all accessors. And tested ;)

**Comment by pscollins** [comment]
I've got it working and tested for every form except 

     obj.NUMBER 

It's listed in the testcases I just pushed, commeted out. Uncommenting it yields:

    test/objects.coffee:474:9: error: unexpected .42
    eq obj.42, 3

I assume that what's happening is `Lexer#numberToken` is consuming the `.` and lexing it as `IDENTIFIER NUMBER` rather than `IDENTIFIER . NUMBER`. 

I'm not sure how this could be dealt with in the parser. The best case I can imagine would be adding in a special `IDENTIFIER . [0-9]+` case in the lexer that comes before `numberToken` is called, but that seems a little hacky. 

It's also worth noting that something like 

    obj = 
      4.1: 1
      4:
        1: 2

will be ambiguous. 

What's the best way to approach it? Make a special case for numeric accessors, and forbid non-integers? Require that numeric accessors go in parens?

**Comment by jashkenas** [comment]
Re: the parser error, I don't know — but as to your second question...

    obj = 
      4.1: 1

... would simply be:

    obj = 
      "4.1": 1

no? I don't see the ambiguity.

**Comment by pscollins** [comment]
@jashkenas Sorry, I didn't mean that defining the object would be ambiguous -- I meant accessing it would be ambiguous. 

My worry is that `obj.4.2` can be interpreted either as `obj[4][2]` or `obj[4.2]`. If your object looks like `obj = {4.2: "foo"}`, you're going to be confused if `obj.4.2` throws an error for trying to access the property `2` on the undefined `obj[4]`. If your object looks like `obj = {4: {2: "foo"}}`, then you're going to be confused if `obj.4.2` gives you back `undefined` for trying to access `obj[4.2]`. 

Or you might have 

    obj = 
      4:
        2: 1
      4.2: 2

Now, what should `obj.4.2` be? 1, or 2? 

It seems to me like it would be easy for someone new to the language to get tripped up by floating point numbers behaving differently from other kinds of numbers, but maybe I'm overreacting?

**Comment by michaelficarra** [comment]
It should be a decimal integer, not floats, not hex, not scientific notation, etc. So my answer is it would be `obj[4][2]`.

File: src/grammar.coffee [hunk 1]
```diff
@@ -258,13 +258,20 @@ grammar =
...
   # or by array index or slice.
   Accessor: [
-    o '.  Identifier',                          -> new Access $2
-    o '?. Identifier',                          -> new Access $2, 'soak'
-    o ':: Identifier',                          -> [LOC(1)(new Access new Literal('prototype')), LOC(2)(new Access $2)]
-    o '?:: Identifier',                         -> [LOC(1)(new Access new Literal('prototype'), 'soak'), LOC(2)(new Access $2)]
+    o '.  AccessorIdentifier',                  -> new Access $2
+    o '?. AccessorIdentifier',                  -> new Access $2, 'soak'
+    o ':: AccessorIdentifier',                  -> [LOC(1)(new Access new Literal('prototype')), LOC(2)(new Access $2)]
+    o '?:: AccessorIdentifierIdentifier',       -> [LOC(1)(new Access new Literal('prototype'), 'soak'), LOC(2)(new Access $2)]
```
Comment (lydell) [inline src/grammar.coffee]: Typo?

Comment (lydell) [inline src/grammar.coffee]: It's correct in the compiled JS though. Remember to rebuild everything.

**Comment by lydell** [comment]
I think `a . b` is valid, so there should be tests for `a . 'abc'` too

**Comment by lydell** [comment]
> and tests that you can call such accesses as well as access deeper properties

You forgot this.,

File: src/grammar.coffee [hunk 2]
```diff
@@ -258,13 +258,20 @@ grammar =
     o 'This'
   ]
 
+  # The symbols that can be used to access an object member:
+  # identifiers, string literals, and numbers.
+  AccessorIdentifier: [
+    o 'Identifier'
+    o 'AlphaNumeric'
```
Comment (lydell) [inline src/grammar.coffee]: A possible solution for the number problem could be to replace this line with:

```coffee
o 'STRING'
o 'INTEGER'
```

Then, in `AlphaNumeric` replace `o 'NUMBER'` with `o 'Number'` and add:

```coffee
Number: [
  o 'NUMBER'
  o 'INTEGER'
]
```

and adjust the `NUMBER` regex in the lexer to match integers separately from other decimal numbers and make `Lexer::numberToken` tag integers as such.

**Comment by michaelficarra** [comment]
I forgot we even allowed leading dots in number literals. Who thought that was a good idea?

**Comment by lydell** [comment]
> I forgot we even allowed leading dots in number literals.

Me too. That’s really making things more complicated here.

**Comment by lydell** [comment]
I think the best way to solve it is to add `return 0 if number.charAt(0) is '.' and @tag() in INDEXABLE` in `Lexer::numberToken`.

**Comment by jashkenas** [comment]
> I forgot we even allowed leading dots in number literals. Who thought that was a good idea?

It's not a good idea ;) We shouldn't have let that sneak in.

**Comment by lydell** [comment]
See #383.

**Comment by jashkenas** [comment]
> It should be a decimal integer, not floats, not hex, not scientific notation, etc. So my answer is it would be obj[4][2].

Agreed.

**Comment by lydell** [comment]
> We shouldn't have let that sneak in.

That makes it sound like they were never intended but there were no tests for it and during some refactor they were allowed by mistake. Since that's not the case, do you think that we should remove them (which makes this PR a lot simpler to finish, but likely results in 10 duplicate issues the same day it is released) or let them stay?

**Comment by jashkenas** [comment]
Ha ha. There's a certain class of feature (of which there are dozens of examples in CoffeeScript) that I intentionally didn't include at the beginning because they're not my personal preferred style ... but that managed to sneak in over time because people ask for them, and you say "why not?". 

This was one of those.

I don't think we can remove them at this point. But perhaps we could tag them specially in the lexer (or something) to make this change easier.

**Comment by michaelficarra** [comment]
Numbers with a leading `.` instead of `0.` are exceptionally uncommon in JS, and I bet even more so in CoffeeScript. Same goes with numbers with trailing `.`. I think we can safely remove them with a simple comment in the changelog.

**Comment by jashkenas** [comment]
I'd love it if that were really the case. My anecdotal sense is that lots of folks prefer to write `speed + 0.5` as `speed + .5`.

Can we employ Github's code search to find out if folks are using it in CoffeeScript?

**Comment by lydell** [comment]
At least in CSS I’ve seen lots of `.5`s. (Don’t know about JS though.)

**Comment by lydell** [comment]
https://github.com/jashkenas/coffeescript/issues/1982#issuecomment-3314044

**Comment by lydell** [comment]
Just found that this would close #1334 as well.

**Comment by shreeve** [comment]
Any status on this PR? Seems like the issue was related to a possible breaking change if some people used numbers like this: ```x = y + .25```. Is breaking this use case the reason for this PR not being accepted?

**Comment by lydell** [comment]
1. It seems like it was decided that disallowing `.5` etc. is too much of a breaking change to be a real option.
2. Some trickery is required to get `foo.1.2` working.
3. There are lots of merge conflicts.

**Comment by pscollins** [comment]
@shreeve IIRC I had trouble getting it to work in the `.5` case and I got busy with other projects --- feel free to revive this/close this PR for a new one/etc.

**Comment by GeoffreyBooth** [comment]
FYI we should probably put this on hold until [tagged template literals](https://github.com/coffeescript6/discuss/issues/28) are supported, unless we’re certain that there won’t be any ambiguity between this proposed syntax and the syntax we’ll need for that. I would think tagged template literals should have a much higher priority if we’re forced to choose between them.

**Comment by GeoffreyBooth** [comment]
@pscollins or anyone on this thread, I’d love to get this updated and targeted against the `2` branch. It seems like it was rather close?

**Comment by GeoffreyBooth** [comment]
What’s going on with this? Is this something we’re still interested in getting finished, or should I close this as an abandoned effort? If we are trying to finish it, are there any breaking changes that would result from merging this in?

**Comment by michaelficarra** [comment]
I can't speak to the quality of this PR, but this is a great feature and should be pursued.

**Comment by GeoffreyBooth** [comment]
At this point the codebase has changed so much since this PR was started that a new PR will be necessary, so I’m closing this one. I’m leaving issue #1334 open as a reminder that we’re interested in pursuing this; that issue has a link to this thread.

**Pull request closed without merging**